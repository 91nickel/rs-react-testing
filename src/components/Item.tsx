import { DeleteButton } from './DeleteButton'

export const TASK_HEADER_MAX_LENGTH = 32
export const TASK_HEADER_DEFAULT_TEXT = 'Здесь должно быть название'

type Props = Task & {
    onDelete: (id: Task['id']) => void;
    onToggle: (id: Task['id']) => void;
};

export const Item = (props: Props) => {

    const header = props.header
        ? props.header.slice(0, TASK_HEADER_MAX_LENGTH)
        : TASK_HEADER_DEFAULT_TEXT

    return (
        <li className="item-wrapper">
            <input
                type="checkbox"
                id={props.id}
                defaultChecked={props.done}
                onChange={() => props.onToggle(props.id)}
            />
            <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
                {props.done ? <s>{header}</s> : header}
            </label>
            <DeleteButton
                disabled={!props.done}
                onClick={() => props.onDelete(props.id)}
            />
        </li>
    )
}
