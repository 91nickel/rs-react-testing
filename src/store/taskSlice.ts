import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { RootState } from './configureStore'

export const ACTIVE_TASKS_LIMIT = 10

export interface taskListState {
    list: Task[];
    notification: string;
    showActiveOnly: boolean;
}

export const initialState: taskListState = {
    list: [],
    notification: '',
    showActiveOnly: false,
}

export function createTaskCloseMessage (header: string) {
    return `Задача "${header}" завершена`
}

export const taskListSlice = createSlice({
    name: 'taskList',
    initialState,
    reducers: {
        setTasks: (state, action: PayloadAction<Task[]>) => {
            state.list = action.payload
        },
        addTask: (state, action: PayloadAction<Task['header']>) => {
            if (state.list.filter(task => !task.done).length < ACTIVE_TASKS_LIMIT) {
                state.list.push({
                    id: crypto.randomUUID(),
                    header: action.payload,
                    done: false,
                })
            }
        },
        completeTask: (state, action: PayloadAction<Task['id']>) => {
            const task = state.list.find((x) => x.id === action.payload)

            if (task) {
                task.done = true
            }
        },
        toggleTask: (state, action: PayloadAction<Task['id']>) => {
            const task = state.list.find((x) => x.id === action.payload)

            if (task) {
                task.done = !task.done

                if (task.done) {
                    state.notification = createTaskCloseMessage(task.header)
                }
            }
        },
        deleteTask: (state, action: PayloadAction<Task['id']>) => {
            state.list = state.list.filter((x) => x.id !== action.payload)
        },
        setNotification: (state, action: PayloadAction<Task['header']>) => {
            state.notification = `Задача "${action.payload}" завершена`
        },
        clearNotification: (state) => {
            state.notification = ''
        },
        toggleActiveOnly: (state) => {
            state.showActiveOnly = !state.showActiveOnly
        },
    },
})

export const {
    addTask,
    completeTask,
    deleteTask,
    toggleTask,
    setTasks,
    setNotification,
    clearNotification,
    toggleActiveOnly,
} = taskListSlice.actions

export default taskListSlice.reducer

export const tasksSelector = (state: RootState) => state.taskList.showActiveOnly
    ? state.taskList.list
        .filter(item => !item.done)
        .slice(0, ACTIVE_TASKS_LIMIT)
    : state.taskList.list
        .filter(item => !item.done)
        .slice(0, ACTIVE_TASKS_LIMIT)
        .concat(state.taskList.list.filter(item => item.done))

export const fullCount = (state: RootState) => state.taskList.list.length

export const completeCount = (state: RootState) =>
    state.taskList.list.filter((x) => x.done).length

export const uncompleteCount = (state: RootState) =>
    state.taskList.list.filter((x) => !x.done).length

export const getNotification = (state: RootState) =>
    state.taskList.notification

export const getShowActiveOnly = (state: RootState) =>
    state.taskList.showActiveOnly
