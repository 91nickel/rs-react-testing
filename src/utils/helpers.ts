import { screen } from '@testing-library/react'

export const MAX_HEADER_LENGTH = 32
export const MIN_HEADER_LENGTH = 1
export const HEADER_MAX_LENGTH_ERROR_MESSAGE = 'Длина заголовка не должна превышать 32 символа'
export const HEADER_MIN_LENGTH_ERROR_MESSAGE = 'Введите заголовок'

export const validateHeaderMax = (text: string): true | string => {
    return text.length <= MAX_HEADER_LENGTH || HEADER_MAX_LENGTH_ERROR_MESSAGE
}

export const validateHeaderMin = (text: string): true | string => {
    return text.length >= MIN_HEADER_LENGTH || HEADER_MIN_LENGTH_ERROR_MESSAGE
}

export const sayHi = () => 'Hello!'

export const headerFieldOptions = {
    maxLength: MAX_HEADER_LENGTH,
    message: HEADER_MAX_LENGTH_ERROR_MESSAGE,
}
