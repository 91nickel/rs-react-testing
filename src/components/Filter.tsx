import { useDispatch, useSelector } from 'react-redux'
import { getShowActiveOnly, toggleActiveOnly } from '../store/taskSlice'

export const Filter = () => {
    const dispatch = useDispatch()
    const value = useSelector(getShowActiveOnly)
    return (
        <>
            <form className="filter" onChange={() => dispatch(toggleActiveOnly())}>
                <label>
                    <input type="checkbox" name="active" defaultChecked={!!value} />
                    <span>{!!value ? 'Все' : 'Активные'}</span>
                </label>
            </form>
        </>
    )
}

export default Filter
