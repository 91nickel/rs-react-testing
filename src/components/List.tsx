import { Item } from './Item'
import { ACTIVE_TASKS_LIMIT } from '../store/taskSlice'

type Props = {
    items: Task[];
    onDelete: (id: Task['id']) => void;
    onToggle: (id: Task['id']) => void;
};

function activeCount(tasks: Task[]): number {
    return tasks.filter(t => !t.done).length
}

export const List = ({items, onDelete, onToggle}: Props) => (
    <ul className="task-list tasks">
        {items
            .reduce((agr, item) => {
                    return item.done || activeCount(agr) < ACTIVE_TASKS_LIMIT
                        ? [...agr, item]
                        : agr
                },
                [] as Task[])
            .map((item) => (
                <Item
                    {...item}
                    key={item.id}
                    onDelete={onDelete}
                    onToggle={onToggle}
                />
            ))}
    </ul>
)
