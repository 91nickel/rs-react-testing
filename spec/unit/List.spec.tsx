import { act, prettyDOM, render, screen } from '@testing-library/react'
import { List } from 'src/components/List'
import { ACTIVE_TASKS_LIMIT, addTask, setTasks, tasksSelector } from 'src/store/taskSlice'
import { getTasksHtml } from '../utils/getTasksHtml'
import { createTasksMock } from '../utils/createTasksMock'
import { App } from 'src/App'
import { store } from 'src/store/configureStore'
import ue from '@testing-library/user-event'

const userEvent = ue.setup({
    advanceTimers: jest.advanceTimersByTime,
})

const mockItems = [
    {
        id: '1',
        header: 'купить хлеб',
        done: false,
    },
    {
        id: '2',
        header: 'купить молоко',
        done: false,
    },
    {
        id: '3',
        header: 'выгулять собаку',
        done: true,
    },
]

describe('Список задач', () => {

    it('отображение списка задач', () => {
        const onDelete = jest.fn()
        const onToggle = jest.fn()

        const items: Task[] = mockItems

        const {rerender, asFragment} = render(
            <List items={items} onDelete={onDelete} onToggle={onToggle}/>,
        )
        const firstRender = asFragment()

        items.pop()

        rerender(<List items={items} onDelete={onDelete} onToggle={onToggle}/>)
        const secondRender = asFragment()

        expect(firstRender).toMatchDiffSnapshot(secondRender)
    })

    it('не больше 10 невыполненных задач в списке', () => {
        const onDelete = jest.fn()
        const onToggle = jest.fn()
        const items = createTasksMock(ACTIVE_TASKS_LIMIT + 1, true)
        const {rerender, asFragment} = render(
            <List items={items} onDelete={onDelete} onToggle={onToggle}/>,
        )
        const tasksList = getTasksHtml()

        // Отображаемое количество задач равно максимальному значению
        expect(tasksList).toHaveLength(ACTIVE_TASKS_LIMIT)
        const firstRender = asFragment()
        items.push(createTasksMock(1, true)[0])
        rerender(<List items={items} onDelete={onDelete} onToggle={onToggle}/>)
        const secondRender = asFragment()

        // Снимок не изменяется при добавлении задач
        expect(firstRender).toMatchDiffSnapshot(secondRender)
    })
})

