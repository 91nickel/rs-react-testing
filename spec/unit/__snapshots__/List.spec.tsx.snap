// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Список задач не больше 10 невыполненных задач 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Список задач не больше 10 невыполненных задач в списке 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;

exports[`Список задач отображение списка задач 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -46,32 +46,7 @@
            alt="Удалить"
            src="[object Object]"
          />
        </button>
      </li>
-     <li
-       class="item-wrapper"
-     >
-       <input
-         checked=""
-         id="3"
-         type="checkbox"
-       />
-       <label
-         for="3"
-       >
-         <s>
-           выгулять собаку
-         </s>
-       </label>
-       <button
-         class="button button-with-icon"
-         data-alt="Удалить"
-       >
-         <img
-           alt="Удалить"
-           src="[object Object]"
-         />
-       </button>
-     </li>
    </ul>
  </DocumentFragment>"
`;
